sympy/matrices/expressions/matexpr.py

```diff
@@ -541,12 +541,19 @@ def recurse_expr(expr, index_ranges={}):
                         d[indices].append(scalar*remove_matelement(elem, *indices))
                         scalar = 1
                 return [(MatrixElement(Add.fromiter(v), *k), k) for k, v in d.items()]
-            elif isinstance(expr, KroneckerDelta):
+            elif isinstance(expr, KroneckerDelta) and all(a in index_ranges for a in expr.args):
```

Comment:
The expression after the `and` is causing the test to fail. It looks like this change is breaking the code, as it assumes that Kronecker deltas have both indices summed over, while it may be the case that one index is not summed.